src/Agent.Sdk/LogPlugin.cs

```diff
@@ -0,0 +1,534 @@
+                tokenSource.Cancel();
+
+                _trace.Trace($"Wait for all plugins finish process outputs.");
+                foreach (var task in processTasks)
+                {
+                    try
+                    {
+                        await task.Value;
+                        _trace.Trace($"Plugin '{task.Key}' finished log process.");
+                    }
+                    catch (Exception ex)
+                    {
+                        _trace.Output($"Plugin '{task.Key}' failed with: {ex}");
+                    }
+                }
+
+                // Stop monitor
+                monitorSource.Cancel();
+                await memoryUsageMonitor;
+
+                // job has finished, all log plugins should start their finalize process
+                Dictionary<string, Task> finalizeTasks = new Dictionary<string, Task>();
+                foreach (var plugin in _plugins)
+                {
+                    string typeName = plugin.GetType().FullName;
+                    if (!_shortCircuited[typeName].Task.IsCompleted)
+                    {
+                        _trace.Trace($"Start finalize for plugin '{plugin.FriendlyName}'");
+                        var finalize = plugin.FinalizeAsync(_pluginContexts[typeName]);
+                        finalizeTasks[plugin.FriendlyName] = finalize;
+                    }
+                    else
+                    {
+                        _trace.Trace($"Skip finalize for short circuited plugin '{plugin.FriendlyName}'");
+                    }
+                }
+
+                _trace.Trace($"Wait for all plugins finish finalization.");
+                foreach (var task in finalizeTasks)
+                {
+                    try
+                    {
+                        await task.Value;
+                        _trace.Trace($"Plugin '{task.Key}' finished job finalize.");
+                    }
+                    catch (Exception ex)
+                    {
+                        _trace.Output($"Plugin '{task.Key}' failed with: {ex}");
+                    }
+                }
+
+                _trace.Trace($"All plugins finished finalization.");
+            }
+        }
+
+        public void EnqueueOutput(string output)
+        {
+            if (!string.IsNullOrEmpty(output))
+            {
+                foreach (var plugin in _plugins)
+                {
+                    string typeName = plugin.GetType().FullName;
+                    if (!_shortCircuited[typeName].Task.IsCompleted)
+                    {
+                        _outputQueue[typeName].Enqueue(output);
+                    }
+                }
+            }
+        }
+
+        public void Finish()
+        {
+            _jobFinished.TrySetResult(0);
+        }
+
+        private async Task StartMemoryUsageMonitor(CancellationToken token)
+        {
+            Dictionary<string, Int32> flag = new Dictionary<string, int>();
```

Comment:
I think this variable is worth a rename. At first glance, it's not clear what `flag` means.

Reply:
sure